### ISSUE TITLE
Reproducing reasonable results on CIFAR-10?

### ISSUE THREAD

**Issue by FutureXiang** [mmathew23/improved_edm#1]
Hi,
Thank you for the great implementation!

I'm also interested in this paper, and I tried to implement it by myself. However, I failed to reproduce reasonable (better than, or as good as the original EDM) generation performance (i.e. FID) on the CIFAR-10 dataset.

I'm really curious to know whether you have successfully achieved promising FID on any unconditional generation dataset.

Thank you!

**Comment by mmathew23** [comment]
Currently working on that (conditional) but could use more compute to go faster. I haven't been able to match EDM's metrics yet, but I also haven't let it train for 200M imgs yet.

Hyperparameter sweeps seem vital for the approach to work and I'm tuning with smaller runs first.

**Comment by mmathew23** [comment]
@FutureXiang As a quick update, in my latest test I get an FID of about 2.4 in conditional cifar 10 after about 50 million images. Not sure how much more it will improve though. 

If you have compute access, would you be willing to help benchmark?

**Comment by FutureXiang** [comment]
Great to hear that!
I will try it later, but could you share more specification of that 2.4 FID model? Does it correspond to the main / config_a branch?

FYI, my implementation of Config G cannot perform well on CIFAR-10 (FID>10), despite the training loss is similar to my EDM1 unconditional implementation, which achieves ~2.2 FID after 60 million images.

I think my conditional EDM1 on CIFAR-10 can reach ~2.12 FID after 40 million images, so your FID is a good start, but it might be further improved.

**Comment by YichengDWu** [comment]
Could you guys share the graph of training loss trajectory? @FutureXiang @mmathew23

**Comment by FutureXiang** [comment]
@YichengDWu 
### The following graph is my training loss curve on unconditional CIFAR-10.
Note that it's another project --- not EDM2. Specifically, I used:
- non-leaky augmentation (proposed by EDM1) at the training;
- 35.7M #params UNet (proposed by DDPM) instead of the Config-A UNet or Config-G UNet.

![image](https://github.com/mmathew23/improved_edm/assets/33350017/92d66b9c-5006-4615-9a3f-91b120f60e5f)

This model achieves 3.3 FID after 2000 epoch training (18 steps deterministic Heun).

### Here is another graph, which represents my failed implementation of Config-G (without post-hoc EMA).
![image](https://github.com/mmathew23/improved_edm/assets/33350017/a6c38f52-0776-4c4f-9ed8-6836b0df06c2)

Despite similar loss trajectories, this model **never** achieves comparable FID (e.g. <4.0).
I think maybe Config-G is very sensitive to EMA rates and the post-hoc EMA is needed.
I'm trying to reproduce the whole progress from Config-A to Config-G to figure it out.

**Comment by YichengDWu** [comment]
@FutureXiang Thanks for sharing, could you explain a bit on why your loss is negative?

**Comment by YichengDWu** [comment]
EDM (2) is designed  so that at initialization, the expectation of loss is 1. My implementation has this property. What I don't know is how good my implementation is trained since I dont have a script to compute the FID yet. 
![W B Chart 1_19_2024, 7_03_12 PM](https://github.com/mmathew23/improved_edm/assets/38997672/29a56749-a8f8-456f-9f85-b1714aa28610)

**Comment by FutureXiang** [comment]
> @FutureXiang Thanks for sharing, could you explain a bit on why your loss is negative?

Because it is the re-weighted version. Please refer to the EDM2 paper (Page 20-22).
If you use the original version, the loss value is expected to be in [0.17, 0.175] after 2000 epochs training.

Btw, note that a reasonable loss value is not sufficient to lead to a good generation quality. Lower doesn't mean the better.

**Comment by YichengDWu** [comment]
I would suggest you turn off the uncertainty net for now. I suspect there is a lot more details to it not mentioned in the paper. 

I'm using Config G, just without the uncertainty net.

**Comment by YichengDWu** [comment]
> the loss value is expected to be in [0.17, 0.175] after 2000 epochs training.

That's good to know. Thanks for the info!

**Comment by mmathew23** [comment]
`accelerate` launch train.py` with the latest changes will run an uncertainty loss based run and `accelerate launch train.py loss_type='scaled'`. I've merged config_a to main, so no need to use that branch.

I've had mixed results with loss_type so far. When I used mlp style, my loss (scaled loss) is actually much better than when I optimize scaled loss directly. But my ema FID score is worse (around 4). I've also run small dataset tests where the mlp version produces better FID than the scaled version. I'm guessing the final results is quite sensitive to the EMA decay rate, so I'm going to implement Post Hoc EMA and see which version produces the better FID.

**Comment by YichengDWu** [comment]
Could you verify if your model is over fitting? I observed that after 500 epochs without data augmentation.

**Comment by mmathew23** [comment]
I've been running a few tests. Indeed, adding the augmentation pipeline from EDM helps for cifar, and the mlp loss weighting does not perform as well as the original loss weightings from the EDM paper. 

Using augmentations and the original EDM loss weights I can get a conditional FID of 1.95 at the end of training with 0.04 sigma_rel. Maybe classifier free guidance could help slightly, and it's possible a previous checkpoint performs better. 

There are a few things on my mind to push even further:

- A comprehensive hyperparameter sweep may show improvements. It's unclear what the best learning rate and t_ref is for the scheduler. 
- The way the paper describes the mlp loss weight feels straightforward but for some reason it's not translating into results yet.
- I've found that a gradient scaler is necessary for fp16 training, even though Karras et al don't use one. 

I'm going to close this for now, as I'm getting a reasonable FID. Feel free to reopen if you have other questions. Or send me a DM on twitter [@DoubleMathew](https://twitter.com/DoubleMathew).
